### PR TITLE
[AAP-75696] feat: add automated PR label workflow

### PR DESCRIPTION
**Pull request by jay-steurer**
## Description

- **What is being changed?** Adds two new GitHub Actions workflows for automated PR status tracking via labels.
- **Why is this change needed?** The jewel repository currently has no automated way to communicate PR status via labels. Team members must manually label PRs to indicate readiness for review, CI failures, or community contributions, which is error-prone and often out of date.
- **How does this change address the issue?** A `pr-labels.yml` workflow automatically applies/removes labels based on CI status, review state, draft status, and author association. A `pr-review-trigger.yml` stub enables the labeler to react to review events on cross-fork PRs.

**Labels managed:** `Ready-for-review`, `fix-ci`, `blocked`, `WIP`, `Review-provided`, `Community`, `Ready-to-merge`

**Security:** This is a public repository. All sensitive data (username mapping, Slack bot token, Slack channel) is stored in GitHub Actions secrets — nothing is hardcoded. The workflow does not check out PR code, eliminating `pull_request_target` attack surface for cross-fork PRs.

**Required GitHub Secrets (admin setup):**

| Secret | Purpose | Required? |
|--------|---------|-----------|
| `USERNAME_MAPPING` | JSON map of GitHub username to Slack member ID | Optional |
| `SLACK_BOT_TOKEN` | Slack Bot OAuth token for notifications | Optional |
| `SLACK_CHANNEL` | Slack channel ID for Ready-to-merge notifications | Optional |

Ported from the validated POC in [platform-services-utilities PR #217](https://github.com/ansible/platform-services-utilities/pull/217).

Jira: AAP-75696

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Self-Review Checklist
- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Prerequisites
- Repository admin access to configure GitHub Actions secrets

### Steps to Test
1. Open a PR against the `devel` branch
2. Verify the `PR Labels` workflow runs on the PR
3. Check that labels are applied based on CI status and PR state
4. Test draft PR → `WIP` label applied
5. Test CI failure → `fix-ci` + `blocked` labels applied
6. Test review submitted → `Review-provided` label applied
7. Optionally configure `SLACK_BOT_TOKEN`, `SLACK_CHANNEL`, and `USERNAME_MAPPING` secrets to test Slack notifications

### Expected Results
- Labels are created automatically if they don't exist
- Labels are applied/removed based on PR events (opened, synchronize, review, draft conversion)
- Slack notification sent when `Ready-to-merge` is first applied (if secrets configured)
- Workflow degrades gracefully when secrets are not configured

### Required Actions
- [ ] Requires infrastructure/deployment changes
  

Assisted-by: Claude Code / Opus 4.6 (Anthropic)

## Summary by CodeRabbit

* **New Features**
  * Automated PR labeling and lifecycle management (fixed label set, auto-create/remove, clears on close/merge and on update).
  * Auto-assigns PR author when unassigned; detects WIP via draft state and title pattern.
  * Computes review and CI readiness; toggles Ready-for-review / Ready-to-merge and manages CI-related labels.
  * Optional Slack notification on newly Ready-to-merge (title escaped; non-fatal if not sent).
  * Optional author classification via configurable mapping.

* **Chores**
  * Added a lightweight workflow to surface review submissions for the labeler.

### CODE REVIEW

**Comment by jay-steurer** [comment]
All four CodeRabbit findings addressed in the latest push:

1. **Pin actions/github-script to SHA** — pinned to `f28e40c7f34bde8b3046d885e986cb6290c5673b` (v7)
2. **Disambiguate fallback PR lookup** — filter now also matches `p.head.repo?.full_name === headRepoFullName` to prevent cross-fork branch name collisions
3. **Guard addAssignees** — pre-check via `GET /repos/{owner}/{repo}/assignees/{assignee}` wrapped in try/catch so failures never block label reconciliation
4. **Zero CI checks treated as pending** — changed to require `ciChecks.length > 0` before marking CI as passed; zero check runs now log and skip

**Comment by tznamena** [comment]
This might be increasing a scope a bit however it would very helpful if we could define the labels outside of the workflow. If it is too much of a scope creep we can handle that in another PR.

File: .github/workflows/pr-labels.yml [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+              const ciChecks = checkRuns.check_runs.filter(
+                cr => cr.name !== 'label'
+              );
+
+              const allCompleted = ciChecks.length > 0
+                && ciChecks.every(cr => cr.status === 'completed');
+
+              if (allCompleted) {
+                const anyFailed = ciChecks.some(cr =>
+                  cr.conclusion === 'failure' || cr.conclusion === 'timed_out'
```
Comment (Dostonbek1) [inline .github/workflows/pr-labels.yml:262]: **Major — Cancelled CI treated as passing.**

`anyFailed` only checks for `failure`/`timed_out`, so `cancelled` or `action_required` conclusions let `ciPassed = true`. A manually cancelled CI run or one cancelled by concurrency limits could result in `Ready-to-merge` being applied without CI actually verifying the code. This fails silently.

```suggestion
                  cr.conclusion === 'failure' || cr.conclusion === 'timed_out'
                  || cr.conclusion === 'cancelled' || cr.conclusion === 'action_required'
```

File: .github/workflows/pr-labels.yml [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+              return;
+            }
+
+            // --- Ensure labels exist ---
+            for (const [name, config] of Object.entries(LABELS)) {
+              try {
+                await github.rest.issues.getLabel({ owner, repo, name });
+              } catch (e) {
+                if (e.status === 404) {
+                  await github.rest.issues.createLabel({
```
Comment (Dostonbek1) [inline .github/workflows/pr-labels.yml]: **Minor — Concurrent label creation race.** Two simultaneous first-run events could both get a 404 from `getLabel` and both call `createLabel`. The second fails with an unhandled 422 ("already_exists"), crashing the workflow.

Wrap `createLabel` in a try/catch that ignores 422:
```javascript
if (e.status === 404) {
  try {
    await github.rest.issues.createLabel({ ... });
    core.info(`Created label: ${name}`);
  } catch (createErr) {
    if (createErr.status !== 422) throw createErr;
    core.info(`Label ${name} already created by another run`);
  }
}
```

File: .github/workflows/pr-labels.yml [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            }
+
+            // --- Reviews (latest state per reviewer) ---
+            const { data: reviews } = await github.rest.pulls.listReviews({
+              owner, repo, pull_number: prNumber,
+            });
+            const latestByReviewer = new Map();
+            for (const r of reviews) {
+              if (['APPROVED', 'CHANGES_REQUESTED'].includes(r.state)) {
+                latestByReviewer.set(r.user.login, r.state);
```
Comment (Dostonbek1) [inline .github/workflows/pr-labels.yml:222]: **Minor — Self-reviews counted toward approval.** The PR author's self-approval counts in `latestByReviewer`, potentially applying `Ready-to-merge` without independent review. Branch protection still gates the actual merge, but the label silently misrepresents the review state.

Consider filtering out the PR author:
```javascript
if (['APPROVED', 'CHANGES_REQUESTED'].includes(r.state) && r.user.login !== pr.user.login) {
```

File: .github/workflows/pr-labels.yml [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+                const normalizedMappingForSlack = Object.fromEntries(
+                  Object.entries(usernameMapping).map(([k, v]) => [k.toLowerCase(), v])
+                );
+                const authorSlackId = normalizedMappingForSlack[normalizedLogin];
+                const mention = authorSlackId ? `<@${authorSlackId}>` : pr.user.login;
+
+                core.setOutput('notify', 'true');
+                core.setOutput('slack_mention', mention);
+                core.setOutput('pr_url', pr.html_url);
+                core.setOutput('pr_title', pr.title);
```
Comment (Dostonbek1) [inline .github/workflows/pr-labels.yml]: **Minor — Defense-in-depth: sanitize PR title before `setOutput`.** The PR title is attacker-controlled in a public repo. While the shell step uses safe env var expansion (not `${{ }}`) and `sed` escapes `<>&`, stripping control characters and truncating would harden this further against edge cases (Slack mrkdwn injection, theoretical GITHUB_OUTPUT delimiter collision).

```suggestion
                const safeTitle = pr.title.replace(/[^\x20-\x7E]/g, '\).slice(0, 200);
                core.setOutput('pr_title', safeTitle);
```

**Review by Dostonbek1 — CHANGES_REQUESTED** [review]
**Code Review: 9.2/10 — NEEDS_CHANGES**

| Lens | Score |
|---|---|
| Functionality | 8.0/10 |
| Security | 9.5/10 |
| Quality | 10.0/10 |

1 Major finding (cancelled CI treated as passing) and 3 Minor findings. See inline comments for details and suggested fixes.

Positives: No code checkout with `pull_request_target`, SHA-pinned action, safe shell env var pattern, well-structured label state machine.

*AI-assisted review — Claude Code / Opus 4.6 (Anthropic)*

File: .github/workflows/pr-labels.yml [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+            } else {
+              await removeLabel('Review-provided');
+            }
+
+            // --- CI status ---
+            let ciPassed = false;
+            const isSynchronize = context.eventName === 'pull_request_target'
+              && context.payload.action === 'synchronize';
+
+            if (isSynchronize) {
```
Comment (bhavenst) [inline .github/workflows/pr-labels.yml:242]: Minor: `checks.listForRef` returns only the first page (default 30). The repo has ~21 checks today so this is fine, but adding `per_page: 100` would future-proof it against missed failures if more checks are added.

**Review by bhavenst — COMMENTED** [review]
Looks good — security model is solid (no code checkout, pinned SHA, graceful secret handling). One minor suggestion on check-run pagination.

**Comment by jay-steurer** [comment]
Addressed all review findings and rebased on latest devel:

**@Dostonbek1** (4 findings):
1. **Cancelled CI as passing** — added `cancelled` and `action_required` to the failure check
2. **Concurrent label creation race** — wrapped `createLabel` in try/catch that ignores 422
3. **Self-reviews counted as approval** — filtered out `pr.user.login` from review tracking
4. **PR title sanitization** — strip non-printable chars and truncate to 200 before `setOutput`

**@bhavenst** (1 finding):
5. **checks.listForRef pagination** — added `per_page: 100`

**@tznamena**: Good idea on externalizing label definitions — agree it's better as a follow-up PR to keep this one focused.

Branch rebased cleanly onto upstream/devel (3 commits).

**Review by bhavenst — APPROVED** [review]
Feedback addressed — per_page:100 added, cancelled/action_required treated as failures, label creation race guarded, author self-reviews excluded, title sanitized. LGTM.

**Review by Dostonbek1 — APPROVED** [review]
**Re-review: 10/10 — APPROVE**

All 4 findings from the prior review (9.2/10) have been addressed in commit ac189ee:

| Finding | Fix |
|---|---|
| Major: Cancelled CI treated as passing | `cancelled` and `action_required` conclusions now included in `anyFailed` check |
| Minor: Label creation race | `createLabel` wrapped in try/catch ignoring 422 |
| Minor: Self-reviews counted | PR author excluded from review tracking via `r.user.login !== pr.user.login` |
| Minor: PR title unsanitized | Control chars stripped and length truncated before `setOutput` |

Bonus: `per_page: 100` added to `checks.listForRef` for future-proofing.

No new issues introduced by the fixes. Clean implementation.

*AI-assisted review — Claude Code / Opus 4.6 (Anthropic)*

**Comment by jay-steurer** [comment]
Addressed new CodeRabbit finding — added a source-event guard to skip `workflow_run` events originating from `push` (e.g., merge to devel). Only `pull_request` and `pull_request_review` source events are processed, preventing the labeler from matching unrelated open PRs by branch name.
